Mathlib/Topology/Compactness/SigmaCompact.lean

```diff
@@ -5,28 +5,169 @@ Authors: Johannes Hölzl, Mario Carneiro, Yury Kudryashov
+  exact isSigmaCompact_iUnion_of_isCompact K.uncurry fun x ↦ (hcomp x.1 x.2)
+
+/-- Countable unions of σ-compact sets are σ-compact. -/
+lemma isSigmaCompact_sUnion (S : Set (Set α)) (hc : Set.Countable S)
+    (hcomp : ∀ s : S, IsSigmaCompact s (α := α)) : IsSigmaCompact (⋃₀ S) := by
+  have : Countable S := countable_coe_iff.mpr hc
+  apply sUnion_eq_iUnion.symm ▸ isSigmaCompact_iUnion _ hcomp
+
+/-- Countable unions of σ-compact sets are σ-compact. -/
+lemma isSigmaCompact_biUnion {ι : Type*} {s : Set ι} {S : ι → Set α} (hc : Set.Countable s)
+    (hcomp : ∀ (i : ι), i ∈ s → IsSigmaCompact (S i)) :
+    IsSigmaCompact (⋃ (i : ι) (_ : i ∈ s), S i) := by
+  have : Countable ↑s := countable_coe_iff.mpr hc
+  rw [biUnion_eq_iUnion]
+  exact isSigmaCompact_iUnion _ (fun ⟨i', hi'⟩ ↦ hcomp i' hi')
+
+/-- A closed subset of a σ-compact set is σ-compact. -/
+lemma IsSigmaCompact.of_isClosed_subset {s t : Set α} (ht : IsSigmaCompact t)
+    (hs : IsClosed s) (h : s ⊆ t) : IsSigmaCompact s := by
+  rcases ht with ⟨K, hcompact, hcov⟩
+  refine ⟨(fun n ↦ s ∩ (K n)), fun n ↦ (hcompact n).inter_left hs, ?_⟩
+  rw [← inter_iUnion, hcov]
+  exact inter_eq_left.mpr h
+
+/-- If `s` is σ-compact and `f` is continuous on a set `w` containing `s`, `f(s)` is σ-compact.-/
+lemma IsSigmaCompact.image_of_continuousOn {f : α → β} {s : Set α} (hs : IsSigmaCompact s)
+    (hf : ContinuousOn f s) : IsSigmaCompact (f '' s) := by
+  rcases hs with ⟨K, hcompact, hcov⟩
+  refine ⟨fun n ↦ f '' K n, ?_, hcov.symm ▸ image_iUnion.symm⟩
+  exact fun n ↦ (hcompact n).image_of_continuousOn (hf.mono (hcov.symm ▸ subset_iUnion K n))
+
+/-- If `s` is σ-compact and `f` continuous, `f(s)` is σ-compact. -/
+lemma IsSigmaCompact.image {f : α → β} (hf : Continuous f) {s : Set α} (hs : IsSigmaCompact s) :
+    IsSigmaCompact (f '' s) := hs.image_of_continuousOn hf.continuousOn
+
+/-- If `f : X → Y` is an `Embedding`, the image `f '' s` of a set `s` is σ-compact
+  if and only `s`` is σ-compact.
+This does not hold for merely inducing maps; direction `←` requires injectivity. -/
+lemma Embedding.isSigmaCompact_iff {f : α → β} {s : Set α}
+    (hf : Embedding f) : IsSigmaCompact s ↔ IsSigmaCompact (f '' s) := by
+  constructor
+  · exact fun h ↦ h.image (continuous hf)
+  · rintro ⟨L, hcomp, hcov⟩
+    -- Suppose f '' s is σ-compact; we want to show f is σ-compact.
+    -- Write f(s) as a union of compact sets L n, so s = ⋃ K n with K n := f⁻¹(L n).
+    -- Since f is an embedding, each K n is compact iff L n is.
+    refine ⟨fun n ↦ f ⁻¹' (L n), ?_, ?_⟩
+    · intro n
+      have : f '' (f ⁻¹' (L n)) = L n := by
+        have h: L n ⊆ f '' s := hcov.symm ▸ subset_iUnion L n
+        apply Set.image_preimage_eq_of_subset (SurjOn.subset_range h)
+      exact hf.toInducing.isCompact_iff.mpr (this.symm ▸ (hcomp n))
+    · calc ⋃ n, f ⁻¹' L n
+        _ = f ⁻¹' (⋃ n, L n) := by rw [preimage_iUnion]
+        _ = f ⁻¹' (f '' s) := by rw [hcov]
+        _ = s := preimage_image_eq s hf.inj
```

Comment:
```suggestion
/-- If `f : X → Y` is `Inducing`, the image `f '' s` of a set `s` is σ-compact
  if and only `s`` is σ-compact. -/
lemma Inducing.isSigmaCompact_iff {f : α → β} {s : Set α}
    (hf : Inducing f) : IsSigmaCompact s ↔ IsSigmaCompact (f '' s) := by
  constructor
  · exact fun h ↦ h.image hf.continuous
  · rintro ⟨L, hcomp, hcov⟩
    -- Suppose f(s) is σ-compact; we want to show s is σ-compact.
    -- Write f(s) as a union of compact sets L n, so s = ⋃ K n with K n := f⁻¹(L n) ∩ s.
    -- Since f is inducing, each K n is compact iff L n is.
    refine ⟨fun n ↦ f ⁻¹' (L n) ∩ s, ?_, ?_⟩
    · intro n
      have : f '' (f ⁻¹' (L n) ∩ s) = L n := by
        rw [image_preimage_inter, inter_eq_left.mpr]
        exact (subset_iUnion _ n).trans hcov.le
      exact hf.isCompact_iff.mpr (this.symm ▸ (hcomp n))
    · calc ⋃ n, f ⁻¹' L n ∩ s
        _ = f ⁻¹' (⋃ n, L n) ∩ s  := by rw [preimage_iUnion, iUnion_inter]
        _ = f ⁻¹' (f '' s) ∩ s := by rw [hcov]
        _ = s := inter_eq_right.mpr (subset_preimage_image _ _)
```
Actually `Inducing` is enough :wink:

Reply:
You can keep the `Embedding` version too of course.

Reply:
Nice find! I've changed the proof as you indicated (and kept the `Embedding` version, following the example for compact sets).